mat64/shadow.go

```diff
@@ -223,3 +220,62 @@ func rectanglesOverlap(off, aCols, bCols, stride int) bool {
 	// b strictly wraps and so must overlap with a.
 	return true
 }
+
+// trianglesOverlap returns whether the strided triangles a and b overlap
+// when b is offset by off elements after a but has at least one element before
+// the end of a. a and b have aSize and bSize respectively.
+func trianglesOverlap(off, aSize, bSize, stride int, aUpper, bUpper bool) bool {
+	if !rectanglesOverlap(off, aSize, bSize, stride) {
+		// Fast return if bounding rectangles do not overlap.
+		return false
+	}
+
+	if aUpper {
+		// Test overlap by projecting diagonals from the b upper
+		// left corner, and upper right or lower left corner onto
+		// the top row of a.
+		rows := off / stride
+		proj := off - rows*(stride+1)
+		if proj >= 0 && proj < aSize {
```

Comment:
This is what I have in mind:
```go
func trianglesOverlap(off, aSize, bSize, stride int, aUpper, bUpper bool) bool {
	if aUpper {
		row := off / stride
		d := off - row*(stride+1) // Signed number of columns (modulo wrapping) between
		                          // b's upper left corner and a's diagonal.
		if d >= 0 && d < aSize - row {
			return true // b's upper left corner lies inside a.
		}
		d += bSize // Shift to b's upper right corner.
		if bUpper && d >= 0 && d < aSize - row {
			return true // b's upper right corner lies inside a.
		}
		return false
	}
	// !aUpper is more complicated.
}
```
Please be skeptical, it might be wrong. And not worth it.